include/hydrogen/SyncInfo.hpp

```diff
@@ -31,12 +36,9 @@ template <Device D> struct SyncInfo
     SyncInfo() {}
 };// struct SyncInfo<D>
 
-template <Device D>
-void AddSynchronizationPoint(SyncInfo<D> syncInfo)
-{}
-
-template <Device D>
-void AddSynchronizationPoint(SyncInfo<D> A, SyncInfo<D> B)
+// Adding synchronization points is generally a no-op
+template <Device... Ds>
+void AddSynchronizationPoint(SyncInfo<Ds>... As)
```

Comment:
Your implementation suggestion is a good one. I don't know that it's clearer, though. I think I'll hold off until we get proper fold support (that should come along in C++17... what? what do you mean it's 2019? ... huh, I'll tell them, but I don't think it'll help... My calendar would like our corner of the world to know it's 2019. We should be able to use C++17 and it should be someone else's problem if the features are not supported.)

As far as exceptions when syncing on different device types, I'm really not sure. You might be saying that you need the device to catch up to the CPU so it can safely transfer values or something. Basically, "AddSyncronizationPoint" should be interpreted as "wait until everyone is caught up to the master". We can discuss further and do something more clever in a future PR.